api/prometheus/v1/remote/remote_api.go

```diff
@@ -0,0 +1,376 @@
+	switch enc {
+	case SnappyBlockCompression:
+		compressed = snappy.Encode(*tmpbuf, inp)
+		if n := snappy.MaxEncodedLen(len(inp)); n > len(*tmpbuf) {
+			// grow the buffer for the next time.
+			*tmpbuf = make([]byte, n)
+		}
+		return compressed, nil
+	default:
+		return compressed, fmt.Errorf("unknown compression scheme [%v]", enc)
+	}
+}
+
+func (r *API) attemptWrite(ctx context.Context, compr Compression, proto WriteProtoFullName, payload []byte, attempt int) (WriteResponseStats, error) {
+	u := r.client.URL("api/v1/write", nil)
+	req, err := http.NewRequest(http.MethodPost, u.String(), bytes.NewReader(payload))
+	if err != nil {
+		// Errors from NewRequest are from unparsable URLs, so are not
+		// recoverable.
+		return WriteResponseStats{}, err
+	}
+
+	req.Header.Add("Content-Encoding", string(compr))
+	req.Header.Set("Content-Type", contentTypeHeader(proto))
+	if proto == WriteProtoFullNameV1 {
+		// Compatibility mode for 1.0.
+		req.Header.Set(versionHeader, version1HeaderValue)
+	} else {
+		req.Header.Set(versionHeader, version20HeaderValue)
+	}
+
+	if attempt > 0 {
+		req.Header.Set("Retry-Attempt", strconv.Itoa(attempt))
+	}
+
+	resp, body, err := r.client.Do(ctx, req)
+	if err != nil {
+		// Errors from Client.Do are likely network errors, so recoverable.
+		return WriteResponseStats{}, retryableError{err, 0}
+	}
+
+	rs, err := parseWriteResponseStats(resp)
+	if err != nil {
+		r.opts.logger.Warn("parsing rw write statistics failed; partial or no stats", "err", err)
+	}
+
+	if resp.StatusCode/100 == 2 {
+		return rs, nil
+	}
+
+	err = fmt.Errorf("server returned HTTP status %s: %s", resp.Status, body)
+	if resp.StatusCode/100 == 5 ||
+		(r.opts.retryOnRateLimit && resp.StatusCode == http.StatusTooManyRequests) {
+		return rs, retryableError{err, retryAfterDuration(resp.Header.Get("Retry-After"))}
+	}
+	return rs, err
+}
+
+// retryAfterDuration returns the duration for the Retry-After header. In case of any errors, it
+// returns 0 as if the header was never supplied.
+func retryAfterDuration(t string) time.Duration {
+	parsedDuration, err := time.Parse(http.TimeFormat, t)
+	if err == nil {
+		return time.Until(parsedDuration)
+	}
+	// The duration can be in seconds.
+	d, err := strconv.Atoi(t)
+	if err != nil {
+		return 0
+	}
+	return time.Duration(d) * time.Second
+}
+
+// writeStorage represents the storage for RemoteWriteHandler.
+// This interface is intentionally private due its experimental state.
+type writeStorage interface {
+	Store(ctx context.Context, proto WriteProtoFullName, serializedRequest []byte) (_ WriteResponseStats, code int, _ error)
+}
+
+type handler struct {
+	logger *slog.Logger
+	store  writeStorage
+}
+
+// NewRemoteWriteHandler returns HTTP handler that receives Remote Write 2.0
+// protocol https://prometheus.io/docs/specs/remote_write_spec_2_0/.
+func NewRemoteWriteHandler(logger *slog.Logger, store writeStorage) http.Handler {
```

Comment:
It's bit inconsistent with NewRemoteAPI in terms of options, let's unify one to one or another 🤔